src/github.com/getlantern/ops/ops.go

```diff
@@ -0,0 +1,147 @@
+// including contextual metadata about the operation and their final success or
+// failure. An op is assumed to have succeeded if by the time of calling Exit()
+// no errors have been reported. The final status can be reported to a metrics
+// facility.
+package ops
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/getlantern/context"
+)
+
+var (
+	cm             = context.NewManager()
+	reporters      []Reporter
+	reportersMutex sync.RWMutex
+)
+
+// Reporter is a function that reports the success or failure of an Op. If
+// failure is nil, the Op can be considered successful.
+type Reporter func(failure error, ctx map[string]interface{})
+
+// Op represents an operation that's being performed. It mimics the API of
+// context.Context.
+type Op interface {
+	// Enter enters a new level on this Op's Context stack.
+	Enter(name string) Op
+
+	// Go starts the given function on a new goroutine.
+	Go(fn func())
+
+	// Exit exits the current level on this Context stack.
+	Exit()
+
+	// Put puts a key->value pair into the current level of the context stack.
+	Put(key string, value interface{}) Op
+
+	// PutDynamic puts a key->value pair into the current level of the context stack
+	// where the value is generated by a function that gets evaluated at every Read.
+	PutDynamic(key string, valueFN func() interface{}) Op
+
+	// Error marks this op as failed if the given err is not nil. If Error is
+	// called multiple times, the latest error will be reported as the failure.
+	// Returns the original error for convenient chaining.
+	Error(err error) error
+
+	// Errorf is like Error but constructs an error from the given message and format args
+	Errorf(msg string, args ...interface{}) error
+}
+
+type op struct {
+	ctx     context.Context
+	failure atomic.Value
+}
+
+// RegisterReporter registers the given reporter.
+func RegisterReporter(reporter Reporter) {
+	reportersMutex.Lock()
+	reporters = append(reporters, reporter)
+	reportersMutex.Unlock()
+}
+
+// Enter enters a new level on the current Op's Context stack, creating a new Op
+// if necessary.
+func Enter(name string) Op {
+	return &op{ctx: cm.Enter().Put("op", name).PutIfAbsent("root_op", name)}
+}
+
+func (o *op) Enter(name string) Op {
+	return &op{ctx: o.ctx.Enter().Put("op", name).PutIfAbsent("root_op", name)}
+}
+
+func (o *op) Go(fn func()) {
+	o.ctx.Go(fn)
+}
+
+// Go mimics the method from context.Manager.
+func Go(fn func()) {
+	cm.Go(fn)
+}
+
+func (o *op) Exit() {
+	var reportersCopy []Reporter
+	reportersMutex.RLock()
+	if len(reporters) > 0 {
+		reportersCopy = make([]Reporter, len(reporters))
+		copy(reportersCopy, reporters)
+	}
+	reportersMutex.RUnlock()
+
+	if len(reportersCopy) > 0 {
+		var failure error
+		_failure := o.failure.Load()
+		ctx := o.ctx.AsMap(_failure, true)
+		if _failure != nil {
+			failure = _failure.(error)
+			_, errorSet := ctx["error"]
+			if !errorSet {
+				ctx["error"] = failure.Error()
+			}
+		}
+		for _, reporter := range reportersCopy {
```

Comment:
Without scanning the code, I never know that `op.Exit` reports failure. We probably need to totally separate those concept entities.